File: euler/euler010.cpp
#include <cstdio>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

typedef unsigned long long ull;

vector<ull> primes;
vector<ull>::iterator it;

bool isPrime(ull n);

int main(int argc, char const *argv[])
{
    int T;
    ull n;
    scanf("%d", &T); // input
    primes.push_back(2); // php: primes[] = 2;
    primes.push_back(3);

    while (T --) {
        ull result = 0;
        scanf("%llu", &n);

        ull i = 0;
        for (; i < primes.size(); i ++) {
            if (primes[i] <= n) {
                result += primes[i];
            } else {
                break;
            }
        }

        if (n <= primes.back()) {
            printf("%llu\n", result);
            continue;
        }

        for (i = primes[i - 1] + 2; i <= n; i ++) {
            if (isPrime(i)) {
                result += i;
            }
        }

        printf("%llu\n", result);
    }

    return 0;
}

bool isPrime(ull n) {
    if (n % 2 == 0) {
        return false;
    }

    for (ull i = 3; i < sqrt(n) + 1; i += 2) {
        if (n % i == 0) {
            return false;
        }
    }

    primes.push_back(n);

    return true;
}

File: euler/euler010.php
<?php

$handle = fopen("php://stdin", "r");
fscanf($handle, "%d", $T);

$primes = [2, 3];

while ($T --) {
    fscanf($handle, "%d", $n);

    solve($n);
}

function solve($n) {
    global $primes;
    $length = count($primes);

    $result = 0;
    $i = 0;
    for (; $i < $length; $i ++) {
        if ($primes[$i] <= $n) {
            $result += $primes[$i];
        } else {
            break;
        }
    }
    if ($primes[$length - 1] >= $n) {
        printf("%d\n", $result);
        return;
    }

    for ($i = $primes[$i - 1] + 2; $i <= $n; $i += 2) {
        if (isPrime($i)) {
            $result += $i;
        }
    }
    printf("%d\n", $result);
}

function isPrime($n) {
    global $primes;

    if ($n == 3) {
        return false;
    }

    for ($i = 3; $i < floor(sqrt($n)) + 2; $i += 2) {
        if ($n % $i == 0) {
            return false;
        }
    }

    $primes[] = $n;

    return true;
}
